pkg/localprovider/localprovider.go

```diff
@@ -0,0 +1,103 @@
+package localprovider
+
+import (
+	"context"
+	"fmt"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	upgradev1alpha1 "github.com/openshift/managed-upgrade-operator/pkg/apis/upgrade/v1alpha1"
+	"github.com/openshift/managed-upgrade-operator/util"
+)
+
+const UPGRADECONFIG_CR_NAME = "managed-upgrade-config"
+
+var log = logf.Log.WithName("upgradeconfig-localprovider")
+
+func New(c client.Client, name string) (*localProvider, error) {
+	return &localProvider{
+		client:  c,
+		cfgname: name,
+	}, nil
+}
+
+type localProvider struct {
+	client  client.Client
+	cfgname string
+}
+
+// Get checks the upgrade config on the cluster with matched name which is going to perform the upgrade
+func (l *localProvider) Get() ([]upgradev1alpha1.UpgradeConfigSpec, error) {
+	log.Info("Read the upgrade config from the cluster directly")
+
+	// Get the current UpgradeConfigs on the cluster
+	instances, err := fetchUpgradeConfigs(l.client)
+	if err != nil {
+		return nil, err
+	}
+
+	// Remove the upgradeConfig is it has been consumed
+	for _, u := range instances.Items {
+		if isUpgradeConfigConsumed(u) {
+			log.Info(fmt.Sprintf("The upgrade for version %s has completed, remove the upgradeConfig.", u.Spec.Desired.Version))
+			err = l.client.Delete(context.TODO(), &u)
```

Comment:
I have two queries for this part. 
1. By default, MUO will reconcile against UpgradeConfig CR named `managed-upgrade-config` as such which means we do not quite care about any other UpgradeConfig. So is the iteration loop required keeping the future work for Hotfix/CVE in mind when there "might" be additional UpgradeConfig instances?
2. Should we actually delete the UpgradeConfig once the upgrade is done? Any harm in keeping it as it is? Would help for reference once upgrade is finished.

Reply:
As Matt mentioned below, it will be deleted by upgrade-config-manager anyway, I will probably remove this part.

Reply:
We can control how long the UpgradeConfigManager decides to keep an UpgradeConfig around post-upgrade via the `watchInterval` config setting in the configmap. ie. if we want it to stick around for a while, we set a watchinterval of days. So that is configurable in the current operator. 

My personal preference is that our various spec providers all behave in the same homogenous way, rather than some of them having some particular behaviour characteristics, and that there is no permanence to the CR itself, and that the UpgradeConfigManager manages that regardless of the upgrade type. I see the UpgradeConfig as analogous to something like a CertificateSigningRequest. But stressing that is just my _personal_ preference. :smile: 

So long story short, as long as the localprovider doesn't return any specs when the UC indicates that the upgrade is complete, it looks good to me.

Reply:
Thanks a lot Matt !! I see your point. So the intent here is to treat UpgradeConfig as ephemeral resources relevant during upgrade time. That makes sense. However, IIUC when the provider is OCM, the necessity to delete the UC is also because of the dependency on UpgradePolicy correct? And for localprovider we do not have any such constraint. My point of view was that if ARO or OCP end user explicitly creates UC and doesn't find it after upgrade if we delete it, we can log it in MUO and also provide documented steps to know upgrade status. Else not delete the UC at all if it's ok to keep it.

Reply:
We can certainly keep the UC around for a very long time post-upgrade with configuration of the `watchInterval` in the configmap (granted, it would still get cleaned up on operator restart).

The clean-up of the UC after upgrade completion on the OSD side isn't really tied to OCM at all or necessary to make that functionality work, it's just the design path we went down for how the operator behaves and treats the resource. (nothing would break if the UC stuck around in OSD land instead of being cleaned up)

We should revisit this closer to the launch date on ARO if it makes sense for how ARO SRE want to interact with the operator, and it will be a simple thing to change - in the meantime I'm keen to keep behaviour as consistent as possible from a maintainability perspective.